paramiko/pipe.py

```diff
@@ -46,25 +47,50 @@ def __init__(self):
         self._closed = False
 
     def close(self):
-        os.close(self._rfd)
-        os.close(self._wfd)
-        # used for unit tests:
         self._closed = True
 
+        # We cannot do anything about closing errors. It is only a
+        # "best effort" approach
+        try:
+            os.close(self._rfd)
+        except Exception:
```

Comment:
Why `OSError` is not used rather than `Exception`?

Reply:
In this case we want to be as broad as possible in the try/except. If any exception is raised during the closing of the rd file descriptor, we still want to guarantee the closing of the wr file descriptor. Catching only one exception type could be too narrow and leave the pipe in an invalid state (one descriptor open and the other one closed)

Reply:
Yes, the problem with catching `OSError` is that if os.close() raises a different exception for whatever unknow reason, the pipe could go into an invalid state and potentially leak opened file descriptors.

Reply:
```py
        try:
            os.close(self._rfd)  # imagine this throws RuntimeError
        except OSError:
            pass
        try:
            os.close(self._wfd)  # in that case, the _wrd would not be closed
        except OSError:
            pass
```

Reply:
I'll leave this `except` as a decision for bitprophet.